bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java

```diff
@@ -1427,14 +1427,34 @@ private static void execStoreOpcodes(WorkerExecutionContext ctx, WorkerData sf,
                 i = operands[0];
                 j = operands[1];
                 k = operands[2];
-                bArray = (BRefValueArray) sf.refRegs[i];
-                if (bArray == null) {
+                BNewArray list = (BNewArray) sf.refRegs[i];
+                if (list == null) {
                     handleNullRefError(ctx);
                     break;
                 }
 
                 try {
-                    bArray.add(sf.longRegs[j], sf.refRegs[k]);
+                    long index = sf.longRegs[j];
+                    BRefType refReg = sf.refRegs[k];
+                    switch (list.getType().getTag()) {
+                        case TypeTags.ARRAY_TAG:
+                            BType elementType = ((BArrayType) list.getType()).getElementType();
+                            if (!checkCast(refReg, elementType)) {
+                                throw BLangExceptionHelper.getRuntimeException(RuntimeErrors.TYPE_MISMATCH,
+                                        elementType, (refReg != null) ? refReg.getType() : BTypes.typeNull);
+                            }
+                            break;
+                        case TypeTags.TUPLE_TAG:
+                            elementType = ((BTupleType) list.getType()).getTupleTypes().get((int) index);
+                            if (!checkCast(refReg, elementType)) {
```

Comment:
I think we bring this checkCast() check to the outside of (end of) the switch case, and make it commin for both the cases.
I think we will get here only for array and tuple types, so we can make that assumption, and ignore the default case (or make ArrayTag the default case).

Reply:
Currently, `any` type is also entering this block. Will fix appropriate places so that only array and tuple will get here